src/client/runner/automation/playback/drag.js

```diff
@@ -1,209 +1,211 @@
-            moveOptions.modifiers     = modifiers;
-            moveOptions.speed         = DRAGGING_SPEED;
-            moveOptions.minMovingTime = MIN_MOVING_TIME;
-            moveOptions.dragMode      = true;
+        return {
+            point:   point,
+            options: options,
+            element: topElement || this.element
+        };
+    }
 
-            var moveAutomation = new MoveAutomation(targetTo, moveOptions);
+    _move ({ element, offsetX, offsetY }) {
+        var moveOptions = new MoveOptions();
 
-            moveAutomation
-                .run()
-                .then(() => window.setTimeout(callback, automationSettings.DRAG_ACTION_STEP_DELAY));
-        },
+        moveOptions.offsetX   = offsetX;
+        moveOptions.offsetY   = offsetY;
+        moveOptions.modifiers = this.modifiers;
 
+        var moveAutomation = new MoveAutomation(element, moveOptions);
 
-        cursorMouseUp: function (callback) {
-            if (pointTo)
-                screenPointTo = positionUtils.offsetToClientCoords(pointTo);
-            else {
-                var offsetPos = positionUtils.getOffsetPosition(el);
+        return moveAutomation
+            .run()
+            .then(() => delay(automationSettings.DRAG_ACTION_STEP_DELAY));
+    }
 
-                screenPointTo = positionUtils.offsetToClientCoords({
-                    x: offsetPos.left,
-                    y: offsetPos.top
-                });
-            }
+    _getEndPoint () {
+        if (domUtils.isDomElement(this.destination))
+            return positionUtils.findCenter(this.destination);
 
-            eventPointTo = automationUtil.getEventOptionCoordinates(el, screenPointTo);
+        var startPoint = mouseUtils.getAutomationPoint(this.element, this.offsetX, this.offsetY);
+        var maxX       = styleUtils.getWidth(document);
+        var maxY       = styleUtils.getHeight(document);
+        var endPoint   = {
+            x: startPoint.x + this.destination.dragOffsetX,
+            y: startPoint.y + this.destination.dragOffsetY
+        };
 
-            if (domUtils.isElementInIframe(el)) {
-                var currentIFrame = domUtils.getIframeByElement(el);
-                if (currentIFrame) {
-                    var screenPointToInIFrame = {
-                        x: screenPointTo.x - styleUtils.getScrollLeft(currentIFrame.contentWindow),
-                        y: screenPointTo.y - styleUtils.getScrollTop(currentIFrame.contentWindow)
-                    };
+        return {
+            x: Math.min(Math.max(0, endPoint.x), maxX),
+            y: Math.min(Math.max(0, endPoint.y), maxY)
+        };
+    }
+
+    _mousedown () {
+        return cursor
+            .leftButtonDown()
+            .then(() => {
+                this.eventArgs = this._calculateEventArguments();
 
-                    topElement = getElementFromPoint(screenPointToInIFrame.x, screenPointToInIFrame.y);
+                if (!this.eventArgs.element) {
```

Comment:
I can't get where this condition is `true`? 
`this._calculateEventArguments().element` is [`topElement || this.element`](https://github.com/DevExpress/testcafe/pull/335/files#diff-f7103dbc2839171e175ce73f92a1d6c2R77).
`this.element` can't be undefined